pkgs/_macro_server/lib/macro_server.dart

```diff
@@ -39,27 +39,73 @@ class MacroServer {
         protocol, service, HostEndpoint(port: serverSocket.port), serverSocket);
   }
 
-  Future<Response> sendToMacro(QualifiedName name, HostRequest request) async {
-    _responseCompleter = Completer<Response>();
-    protocol.send(_lastSocket!.add, request.node);
-    return _responseCompleter!.future;
+  /// Sends to the macro identified by `request#macroAnnotation`.
+  ///
+  /// If no such macro is connected, repeatedly waits [timeout] for a new
+  /// macro to connect and checks again.
+  ///
+  /// Throws [TimeoutException] if no such macro connects in the time allowed.
+  Future<Response> sendToMacro(HostRequest request,
+      {Duration timeout = const Duration(seconds: 5)}) async {
+    _Connection? connection;
+    while (true) {
+      // Look up the connection with the macro corresponding to `annotation`.
+      connection = _connections
+          .where((c) => c.descriptions.any(
+              (d) => d.annotation.string == request.macroAnnotation.string))
+          .singleOrNull;
+      // If it's found: done.
+      if (connection != null) break;
+      // Not found, wait [timeout] then recheck. Throws `StateError` on
+      // timeout.
+      await _macroDescriptionBecomesKnown.stream.first.timeout(timeout);
```

Comment:
How do we know this is the correct one?

Reply:
We don't; execution continues from line 53 which returns if the matching macro is now found or waits again if it's not.